homeassistant/components/xiaomi_miio/config_flow.py

```diff
@@ -47,36 +51,69 @@ async def async_step_user(self, user_input=None):
             step_id="user", data_schema=CONFIG_SCHEMA, errors=errors
         )
 
+    async def async_step_zeroconf(self, discovery_info):
+        """Handle zeroconf discovery."""
+        name = discovery_info.get("name")
+        self.host = discovery_info.get("host")
+        mac_address = discovery_info.get("properties", {}).get("mac")
+
+        if not name or not self.host or not mac_address:
+            return self.async_abort(reason="not_xiaomi_miio")
+
+        # Check which device is discovered.
+        if name.startswith(ZEROCONF_GATEWAY):
+            unique_id = format_mac(mac_address)
+            await self.async_set_unique_id(unique_id)
+            self._abort_if_unique_id_configured({CONF_HOST: self.host})
+
+            return await self.async_step_gateway()
+
+        # Discovered device is not yet supported
+        _LOGGER.debug(
+            "Not yet supported Xiaomi Miio device '%s' discovered with host %s",
+            name,
+            self.host,
+        )
+        return self.async_abort(reason="not_xiaomi_miio")
+
     async def async_step_gateway(self, user_input=None):
         """Handle a flow initialized by the user to configure a gateway."""
         errors = {}
         if user_input is not None:
-            host = user_input[CONF_HOST]
             token = user_input[CONF_TOKEN]
+            if user_input.get(CONF_HOST):
+                self.host = user_input[CONF_HOST]
 
             # Try to connect to a Xiaomi Gateway.
             connect_gateway_class = ConnectXiaomiGateway(self.hass)
-            await connect_gateway_class.async_connect_gateway(host, token)
+            await connect_gateway_class.async_connect_gateway(self.host, token)
             gateway_info = connect_gateway_class.gateway_info
 
             if gateway_info is not None:
-                unique_id = f"{gateway_info.model}-{gateway_info.mac_address}-gateway"
+                mac = format_mac(gateway_info.mac_address)
+                unique_id = mac
                 await self.async_set_unique_id(unique_id)
                 self._abort_if_unique_id_configured()
                 return self.async_create_entry(
                     title=user_input[CONF_NAME],
                     data={
                         CONF_FLOW_TYPE: CONF_GATEWAY,
-                        CONF_HOST: host,
+                        CONF_HOST: self.host,
                         CONF_TOKEN: token,
-                        "gateway_id": unique_id,
                         "model": gateway_info.model,
-                        "mac": gateway_info.mac_address,
+                        "mac": mac,
                     },
                 )
 
             errors["base"] = "connect_error"
 
+        if self.host:
+            return self.async_show_form(
```

Comment:
We could set the `data_schema` variable here instead and have a common call to `self.async_show_form` below.

Reply:
done